libs/muddle/src/punishment_broadcast_channel.cpp

```diff
@@ -0,0 +1,252 @@
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "muddle/punishment_broadcast_channel.hpp"
+
+using fetch::muddle::PunishmentBroadcastChannel;
+using fetch::muddle::QuestionStruct;
+
+QuestionStruct PunishmentBroadcastChannel::AllowPeerPull()
+{
+  FETCH_LOCK(lock_);
+  return question_;
+}
+
+PunishmentBroadcastChannel::PunishmentBroadcastChannel(Endpoint &endpoint, MuddleAddress address,
+                                                       CallbackFunction call_back,
+                                                       CertificatePtr certificate, uint16_t channel,
+                                                       bool ordered_delivery)
+  : endpoint_{endpoint}
+  , address_{std::move(address)}
+  , deliver_msg_callback_{std::move(call_back)}
+  , channel_{channel}
+  , certificate_{std::move(certificate)}
+  , rpc_client_{"PunishmentBC", endpoint_, SERVICE_PBC, channel_}
+  , state_machine_{std::make_shared<StateMachine>("PBCStateMach", State::INIT)}
+{
+  FETCH_UNUSED(ordered_delivery);
+  this->Expose(PULL_INFO_FROM_PEER, this, &PunishmentBroadcastChannel::AllowPeerPull);
+
+  // TODO(HUT): rpc beacon rename.
+  // Attaching the protocol
+  rpc_server_ = std::make_shared<Server>(endpoint_, SERVICE_PBC, channel_);
+  rpc_server_->Add(RPC_BEACON, this);
+
+  // Connect states to the state machine
+  state_machine_->RegisterHandler(State::INIT, this, &PunishmentBroadcastChannel::OnInit);
+  state_machine_->RegisterHandler(State::RESOLVE_PROMISES, this,
+                                  &PunishmentBroadcastChannel::OnResolvePromises);
+}
+
+PunishmentBroadcastChannel::~PunishmentBroadcastChannel() = default;
+//{
+//  // TODO(HUT): reinstate this fix.
+//  /* rpc_server_->Remove(RPC_BEACON); */
+//}
+
+// Interface methods
+bool PunishmentBroadcastChannel::ResetCabinet(CabinetMembers const &cabinet)
+{
+  FETCH_LOCK(lock_);
+  current_cabinet_ = cabinet;
+
+  // Set threshold depending on size of committee
+  if (current_cabinet_.size() % 3 == 0)
+  {
+    threshold_ = static_cast<uint32_t>(current_cabinet_.size() / 3 - 1);
+  }
+  else
+  {
+    threshold_ = static_cast<uint32_t>(current_cabinet_.size() / 3);
+  }
+
+  assert(current_cabinet_.size() > 3 * threshold_);
+
+  return true;
+}
+
+/**
+ * Set/reset the question, saving the old question, so that peer who have not yet
+ * reset can still access this information
+ *
+ */
+void PunishmentBroadcastChannel::SetQuestion(ConstByteArray const &question,
```

Comment:
Perhaps it makes sense to pass QA by-value here, this method is called with rvalues in the next overload.

Reply:
I think the linter complains about this